Add CVE-2024-49138 (#42)
googleprojectzero/0days-in-the-wild: 0day-RCAs/2024/CVE-2024-49138.md

# CVE-2024-49138: Windows Common Log File System Driver Elevation of Privilege Vulnerability
*Ong How Chong (STAR Labs SG Pte. Ltd.)*

## The Basics

**Disclosure or Patch Date:** December 10, 2024

**Product:** Windows

**Advisory:** https://msrc.microsoft.com/update-guide/vulnerability/CVE-2024-49138

**Affected Versions:** Before security updates of December 10, 2024, for Windows 10, 11 and Windows Server 2008, 2012, 2016, 2019, 2022, 2025

**First Patched Version:** Security updates of December 10, 2024, for CVE-2024-49138

**Issue/Bug Report:** N/A

**Patch CL:** N/A

**Bug-Introducing CL:** N/A

**Reporter(s):** Advanced Research Team with CrowdStrike

## The Code

**Proof-of-concept:** https://github.com/MrAle98/CVE-2024-49138-POC

**Exploit sample:** See PoC

**Did you have access to the exploit sample when doing the analysis?** Yes

## The Vulnerability

**Bug class:** Untrusted Pointer Dereference (CWE-822)

> **Vulnerability details:** CVE-2024-49138 is a logical vulnerability within the Windows Common Log File System kernel driver that occurs when processing maliciously formatted log files. This will eventually lead to an untrusted pointer dereference by the kernel. An attacker is able to modify the pointer such that the the execution flow of the kernel will be redirected to a user controlled address.

Reviewer: ```suggestion
**Vulnerability details:** CVE-2024-49138 is a logical vulnerability within the Windows Common Log File System kernel driver that occurs when processing maliciously formatted log files. This will eventually lead to an untrusted pointer dereference by the kernel. An attacker is able to modify the pointer such that the execution flow of the kernel will be redirected to a user controlled address.
```

The Windows Common Log File System (CLFS) uses Base Log Files (BLF) and Container files for storing log information. BLFs hold all the important metadata relating to the log file while Containers hold the actual log data.

The untrusted pointer dereference can be found in the function `CClfsBaseFilePersisted::LoadContainerQ()`, where `pContainer` is a variable that can be indirectly manipulated and dereferenced.

```
CClfsBaseFilePersisted::LoadContainerQ(){
	...
    return_value = CClfsBaseFilePersisted::FlushImage((PERESOURCE *)this);
    ...
    if ( return_value < 0 )
        goto LABEL_116;
	...
	LABEL_116:
        ContainerContext->pContainer->Release(ContainerContext->pContainer);	//pContainer dereference
}
```

> In order for `pContainer` dereference to be reached, `CClfsBaseFilePersisted::FlushImage()` would have to return a negative value, or an error code. Provided is the pesudocode for the relevant functions.

Reviewer: ```suggestion
In order for `pContainer` dereference to be reached, `CClfsBaseFilePersisted::FlushImage()` would have to return a negative value, or an error code. Provided is the pseudocode for the relevant functions.
```

* A sector signature and `pContainer` overlap.
* A sector signature and the `signatures array` overlap.

We also need to ensure that `ullDumpCount`, a value found in the BLF header, is an odd value as we want `Usn` to be incremented in `CClfsBaseFilePersisted::WriteMetadataBlock()`. `Usn` will then be used in `ClfsEncodeBlock()` to calculate a new sector signature.

`ClfsEncodeBlock()` calls `ClfsEncodeBlockPrivate()`.

```
ClfsEncodeBlockPrivate(...){
    if ( size or offsets are wrong ){    //preliminary checks
        return ErrorCode;
    }

    SectorNumber = 0;
    while (SectorNumber < TotalSectors){
        //calculate new sector signature
        ...
	    
        SectorSigOffset = SectorNumber << 9;    //SectorNumber * 512 (size of sector)
        *SignaturesArray = *(SectorSigOffset + 2);    //store old signature
        SignaturesArray += 2;    //move to next entry in SignaturesArray
        *(SectorSigOffset + 2) = NewSig    //write new signature
        SectorNumber++;
    }
}
```
ClfsEncodeBlockPrivate() logic:
* `ClfsEncodeBlock()` and `ClfsEncodeBlockPrivate()` does preliminary checks that various fields in the BLF are valid.

> * Loops over all sector signatures and copy them into the `signatures array` located at `SignaturesOffset`, then calculates and writes a new signature value into sector signature (going from low address to high address).

Reviewer: The wording "Loops over all sector signatures" is a bit confusing in my opinion.

```suggestion
* Loops over every 2 data bytes located at sector signature locations and copies them into the `signatures array` located at `SignaturesOffset`, then calculates and writes a signature value into the sector signature location (going from low address to high address).
```

> * As an example, take a malicious BLF which has a `ullDumpCount` value of 2 and a `Usn` value of 1. Following the sector signature format of `[Sector Block Type][Usn]`, each sector of this BLF would have a sector signature of `0x10 0x01`.

Reviewer: This doesn't feel like a continuation of previous bulllets.

```suggestion

As an example, take a malicious BLF which has a `ullDumpCount` value of 2 and a `Usn` value of 1. Following the sector signature format of `[Sector Block Type][Usn]`, each sector of this BLF would have a sector signature of `0x10 0x01` (except for the first and last sector in a block, which have [extra flags](https://github.com/ionescu007/clfs-docs?tab=readme-ov-file#sector-signatures)).
```

* `WriteMetadataBlock()` will increment both `ullDumpCount` and `Usn` by 1 before jumping to `ClfsEncodeBlock()` and `ClfsEncodeBlockPrivate()`.
* `ClfsEncodeBlockPrivate()` would calculate the new sector signature as `0x10 0x02`, following the sector signature format.

> * Each sector would have its old signature of `0x10 0x01` stored at the `signatures array` located at `SignaturesOffset`, and its new sector signature value of `0x10 0x02` be written to the sector signature. However, due to the overlapping header sections that are present in the maliciou BLF, when `ClfsEncodeBlockPrivate()` stores the old signature into the `signatures array`, it will overwrite the new sector signataure value of that sector.

Reviewer: I think this is incorrect. The `signatures array` does not contain the old signatures, but rather the original data. My understanding of what happens is rather that:

* Since the signature array overlaps with a sector boundary, **one of the entries** in the signature array will contain the initial signature of that sector, i.e. `0x10 0x01`. All the other entries will contain `0x00 0x00`.
* When the BLF is loaded, the entries of the signature array are patched in the respective signature slots.
* When `ClfsEncodeBlockPrivate()` iterates over the sectors, it:
  * First copies out the two data bytes from the signature slot to the signature array
  * Then writes the signature (`0x10 0x02`) in the signature slot
 
Let's take example of [Iandoli's blog](https://security.humanativaspa.it/cve-2024-49138-windows-clfs-heap-based-buffer-overflow-analysis-part-1/):

* pContainer overlaps with signature of sector[10]
* signature_array[59] overlaps with signature of sector[11], which contains initial sector signature `0x10 0x01`

When `ClfsEncodeBlockPrivate()` iterates over the sectors, the following sequence happens:

* (Update signatures of sector 0 to 9, not important)
* Update signature of sector 10. Part of `pContainer` gets corrupted with `0x10 0x02`
* Update signature of sector 11. signature_array[59] gets overwritten with `0x10 0x02` (but not important because another overwrite will happen in later step).
* (Update signatures of sector 10 to 58, not important)
* Update signature of sector 59. The original bytes of signature slot in sector 59 (`0x10 0x01`) are placed in signature_array[59], (with overlaps with signature slot of sector[11] !). Signature of sector 59 is set to `0x10 0x02`.
* (Update signatures of sector 60 to 61)

End result:

* pContainer is changed
* sector[11] has invalid signature of `0x10 0x01`

> * As an example:

Reviewer: ```suggestion

As an example:
```

Reviewer: As described in my verbose comment above, I think there is a bit more nuance to it than described here.

>     * Assume the malicious BLF has its `signatures array` located inbetween sectors 1 and 2, meaning that 2 bytes of the `signatures array` would overlap with the sector signature of sector 1. For illustration purposes, assume that the 2 bytes overlapping with the sector signature are the 19th and 20th byte of `signatures array` (10th signature).

Reviewer: ```suggestion
    * Assume the malicious BLF has its `signatures array` located in between sectors 1 and 2, meaning that 2 bytes of the `signatures array` would overlap with the sector signature of sector 1. For illustration purposes, assume that the 2 bytes overlapping with the sector signature are the 19th and 20th byte of `signatures array` (10th signature).
```

With this, an error code is returned all the way up to `LoadContainerQ()`, and we are able to dereference `pContainer`, which would contain a user controlled address.

**Patch analysis:**

Before the patch:
```
CClfsBaseFilePersisted::LoadContainerQ(){
	...
    return_value = CClfsBaseFilePersisted::FlushImage((PERESOURCE *)this);
    ...
	if ( return_value < 0 )
        goto LABEL_116;
...
	LABEL_116:
		ContainerContext->pContainer->Release(ContainerContext->pContainer);	//pContainer dereference
}
```

After the patch:
```
CClfsBaseFilePersisted::LoadContainerQ(){
	...
    if (patch_flag)
        v58 = ContainerContext->pContainer;    //saves pContainer here
	...
    return_value = CClfsBaseFilePersisted::FlushImage((PERESOURCE *)this);
    ...
    if ( return_value < 0 )
        goto LABEL_116;
}
...
LABEL_116:
          v58->Release(v58);	//safe pContainer dereference
```

After the patch, the address of `pContainer` is saved into `v58` before `CClfsBaseFilePersisted::FlushImage()` is called, and subsequently loaded before it is dereferenced. Thus, `pContainer` can no longer be tampered with before it is dereferenced.

> **Thoughts on how this vuln might have been found _(fuzzing, code auditing, variant analysis, etc.)_:** The CLFS driver had been patched for Elevation of Privilege vulnerabilities numerous times over the years. Additionally, the reporter of this CVE was crowdstrike, who detected the vulnerability actively exploited by threat actors. There is a good chance that this was developed by threat actors who understand that the CLFS driver has many vectors to exploit as seen from past patches.

Reviewer: ```suggestion
**Thoughts on how this vuln might have been found _(fuzzing, code auditing, variant analysis, etc.)_:** The CLFS driver had been patched for Elevation of Privilege vulnerabilities numerous times over the years. Additionally, the reporter of this CVE was Crowdstrike, who detected the vulnerability actively exploited by threat actors. There is a good chance that this was developed by threat actors who understand that the CLFS driver has many vectors to exploit as seen from past patches.
```

**(Historical/present/future) context of bug:** 

* 10 December, 2024: CVE-2024-49138 is patched and publicly disclosed.
* 29 January, 2025: Alessandro Iandoli publishes a detailed writeup and POC for CVE-2024-49138.

## The Exploit

(The terms *exploit primitive*, *exploit strategy*, *exploit technique*, and *exploit flow* are [defined here](https://googleprojectzero.blogspot.com/2020/06/a-survey-of-recent-ios-kernel-exploits.html).)

> **Exploit strategy (or strategies):** Modify `pContainer` so that the kernel will dereference it and jump to a user controlled address, leading to unintended behaviour. We can chain this with a different exploit that requires highjacking the execution flow of the kernel in order to achieve elevation of privileges.

Reviewer: ```suggestion
**Exploit strategy (or strategies):** Modify `pContainer` so that the kernel will dereference it and call a chosen kernel function from a usermode vtable.
```

`CClfsBaseFilePersisted::LoadContainerQ()`, which contains the code that dereferences `pContainer`, can be reached by calling the user exposed Common Log File System (CLFS) API `CreateLogFile()`.

**Exploit flow:**

--- POC execution ---
* Create and load malicious BLF file.
* Create a fake `CClfsContainer` object at a user controlled address with a fake vtable that points to the address of `nt!PoFxProcessorNotification`.
* Write additional data in the memory region allocated to the fake `CClfsContainer` object such as the address of `nt!DbgkpTriageDumpRestoreState` and the address of `_KTHREAD.PreviousMode` of the current thread.
* Call `CreateLogFile()` which opens the malicious BLF file.

--- Driver execution ---
* Doing the above dereferences the malicious `CClfsContainer` object at a user controlled address.
* This calls `nt!PoFxProcessorNotification` which redirects the execution flow to `nt!DbgkpTriageDumpRestoreState` which is used to obtain arbitrary 8 byte write.

> * Use this write to overwrite the `_KTHREAD.PreviousMode` of the current thread to 0 (kernel permissions), granting us arbitrary read/write primitives.

Reviewer: ```suggestion
* Use this primitive to overwrite the `_KTHREAD.PreviousMode` of the current thread to 0 (kernelmode), granting us arbitrary read/write primitives to the whole address space using NtReadVirtualMemory() and NtWriteVirtualMemory().
```

--- POC execution ---

> * POC can now read and copy the systems `_EPROCESS.Token` using a series of call to `NtReadVirtualMemory()/NtWriteVirtualMemory()` and plant it into our current process, giving our user mode POC the same privileges as system.

Reviewer: ```suggestion
* POC can now read and copy the system `_EPROCESS.Token` using a series of call to `NtReadVirtualMemory()/NtWriteVirtualMemory()` and plant it into our current process, giving our user mode POC the same privileges as the system process.
```

* Spawn a cmd shell with system privileges.

> In depth analysis of the chained exploit has been performed and documented below.

Reviewer: ```suggestion
In depth analysis of the chained kernel functions has been performed and documented below.
```

mov     rax, qword ptr [rax+68h]    //moves (kernel address + nt!DbgkpTriageDumpRestoreState offset) into rax
mov     rcx, qword ptr [rcx+48h]    //moves start address of arguments into rcx
call    cs:__guard_dispatch_icall_fptr
...
jmp     rax    //jump to nt!DbgkpTriageDumpRestoreState
...
```

nt!DbgkpTriageDumpRestoreState:

At this point, `rcx` contains the address of the arguments being passed.
```
...
DbgkpTriageDumpRestoreState proc near
mov     eax, [rcx+0Ch]
mov     rdx, [rcx]         //rdx is (AddressOfUserThread + offset_KTHREAD.PreviousMode - 0x2078)
mov     [rcx+18h], eax
mov     eax, [rcx+10h]     //eax is first 4 bytes of arbitrary write value
mov     [rdx+2078h], eax   //write to (AddressOfUserThread + offset_KTHREAD.PreviousMode)
mov     rdx, [rcx]
mov     eax, [rcx+14h]     //eax is next 4 bytes of arbitrary write value
mov     [rdx+207Ch], eax   //write to (AddressOfUserThread + offset_KTHREAD.PreviousMode + 4h)
retn
DbgkpTriageDumpRestoreState endp
...
```

**Known cases of the same exploit flow:** There have been other exploits that all have the end-goal of being able to read and copy the systems `_EPROCESS.Token`, thus leading to elevation of privilege.

> **Part of an exploit chain?** Was likely used as part of an exploit chain.

Reviewer: I think it was used standalone to elevate privileges on a Windows machine?

## The Next Steps

### Variant analysis

**Areas/approach for variant analysis (and why):** N/A

**Found variants:** N/A

### Structural improvements

What are structural improvements such as ways to kill the bug class, prevent the introduction of this vulnerability, mitigate the exploit flow, make this type of vulnerability harder to exploit, etc.?

> **Ideas to kill the bug class:** N/A

Reviewer: We can mention this: https://techcommunity.microsoft.com/blog/microsoft-security-blog/security-mitigation-for-the-common-log-filesystem-clfs/4224041

> **Ideas to mitigate the exploit flow:** N/A

Reviewer: SMAP would kill the usermode dereference, but has been discarded on Windows (https://github.com/microsoft/MSRC-Security-Research/blob/master/papers/2020/Evaluating%20the%20feasibility%20of%20enabling%20SMAP%20for%20the%20Windows%20kernel.pdf)

The leaks via NtQuerySystemInformation from medium IL are killed on Windows 11 24H2.

Overwriting Previousmode should also not work anymore on Windows 11 24H2.

See "Limitations and improvements" at https://security.humanativaspa.it/cve-2024-49138-windows-clfs-heap-based-buffer-overflow-analysis-part-1/

**Other potential improvements:** N/A

### 0-day detection methods

> What are potential detection methods for similar 0-days? Meaning are there any ideas of how this exploit or similar exploits could be detected **as a 0-day**? N/A

Reviewer: We can mention:

* Dropping and modifying BLF files
* Spawning cmd.exe as SYSTEM
* Monitoring suspicious API calls such as NtQuerySystemInformation
